public/utils.ts

```diff
@@ -0,0 +1,262 @@
+    } else {
+      if (current_container instanceof Map) {
+        if (
+          !current_container.has(currentKey) ||
+          !(current_container.get(currentKey) instanceof Map)
+        ) {
+          current_container.set(currentKey, new Map<any, any>());
+        }
+        current_container = current_container.get(currentKey);
+      } else {
+        if (
+          !current_container[currentKey] ||
+          typeof current_container[currentKey] !== "object"
+        ) {
+          (current_container as { [key: string]: any })[currentKey] = {};
+        }
+        current_container = current_container[currentKey];
+      }
+    }
+  }
+
+  return container;
+};
+
+export const generateRangeFromStep = (
+  start: number,
+  end: number,
+  step: number,
+): number[] => {
+  if (step <= 0) {
+    throw new Error("Step must be greater than 0");
+  }
+  if (start > end && step > 0) {
+    throw new Error("Start cannot be greater than end when step is positive");
+  }
+
+  let returnArray: number[] = [];
+  let track = start;
+  while (track <= end) {
+    returnArray.push(track);
+    track = roundTwoDecimals(track + step);
+  }
+
+  return returnArray;
+};
+
+export const generateRangeFromSlots = (
+  start: number,
+  end: number,
+  slots: number,
+): number[] => {
+  let returnArray: number[] = [];
+
+  if (slots === 0) {
+    return returnArray;
+  }
+  if (slots === 1) {
+    return [start, end];
+  }
+  if (start === end) {
+    return [start];
+  }
+  if (start > end) {
+    [start, end] = [end, start];
+  }
+
+  let step = (end - start) / slots;
+
+  for (let i = start; i <= end; i += step) {
+    returnArray.push(roundTwoDecimals(i));
+  }
+
+  return returnArray;
+};
+
+export const distinct = (arr: any[]): any[] => {
+  // @ts-ignore
+  return [...new Set<any>(arr)];
+};
+
+export const roundTwoDecimals = (x: number) => Math.round(x * 100) / 100;
+
+// flattenDict is used to flatten a nested object into a single level object
+// Turns {a: {b: {c: 1, d: 2}}} into {'a.b.c': 1, 'a.b.d': 2}
+export const flattenDict = (obj: any | any[], parentKey = "", result = {}) => {
+  const objIsArray = Array.isArray(obj);
+
+  for (const [key, value] of Object.entries(obj)) {
+    const storeKey = objIsArray ? `[${key}]` : key;
+    let newKey = parentKey ? `${parentKey}.${storeKey}` : storeKey;
+
+    if (value && (Array.isArray(value) || typeof value === "object")) {
+      flattenDict(value, newKey, result);
+    } else {
+      // @ts-ignore
+      result[newKey] = value;
+    }
+  }
+  return result;
+};
+
+// objStringifyInOrder is used to stringify objects in a consistent order, flattening nested objects then sorting keys
+export const objStringifyInOrder = (obj: any): string => {
+  let flattenedObj = flattenDict(obj);
+
+  let returnString = "";
+
+  Object.keys(flattenedObj)
+    .sort()
+    .forEach((key) => {
+      // @ts-ignore
+      returnString += `${key}: ${flattenedObj[key]}\n`;
+    });
+
+  return returnString;
+};
```

Comment:
Maybe a better way to do this - doesn't require flattening the object firs https://stackoverflow.com/a/53593328

```
function JSONstringifyOrder(obj, space)
{
    const allKeys = new Set();
    JSON.stringify(obj, (key, value) => (allKeys.add(key), value));
    return JSON.stringify(obj, Array.from(allKeys).sort(), space);
}
```

Reply:
Somewhat addressed this in https://github.com/zengm-games/facesjs/pull/35#discussion_r1559782144

Reply:
I'm don't think that comment changes anything with respect to the specific implementation of stringifying JSON while accounting for the order of object keys.